File: backend/models/ChatMessage.js
const prisma = require('../lib/prisma');
const { v4: uuidv4 } = require('uuid');

class ChatMessage {
  constructor(data) {
    this.id = data.id || uuidv4();
    this.content = data.content;
    this.type = data.type; // 'user' or 'assistant'
    this.timestamp = data.timestamp || new Date();
    this.context = data.context;
    this.isError = data.isError || false;
    this.isDiagnostic = data.isDiagnostic || false;
    this.diagnosticQuestions = data.diagnosticQuestions;
    this.diagnosticAnswers = data.diagnosticAnswers;
    this.parentMessageId = data.parentMessageId;
    this.sessionId = data.sessionId;
    this.createdAt = data.createdAt || new Date();
  }

  async save() {
    const data = {
      id: this.id,
      content: this.content,
      type: this.type,
      timestamp: this.timestamp,
      context: this.context,
      isError: this.isError,
      isDiagnostic: this.isDiagnostic,
      diagnosticQuestions: this.diagnosticQuestions,
      diagnosticAnswers: this.diagnosticAnswers,
      parentMessageId: this.parentMessageId,
      sessionId: this.sessionId,
    };

    // Always create new message (since we auto-generate IDs)
    const created = await prisma.chatMessage.create({
      data: data,
    });
    return new ChatMessage(created);
  }

  static async getAll() {
    const messages = await prisma.chatMessage.findMany({
      orderBy: { timestamp: 'asc' },
    });
    
    return messages.map(message => new ChatMessage(message));
  }

  static async getById(id) {
    const message = await prisma.chatMessage.findUnique({
      where: { id },
    });
    
    if (message) {
      return new ChatMessage(message);
    }
    
    return null;
  }

  static async getRecent(limit = 50) {
    const messages = await prisma.chatMessage.findMany({
      orderBy: { timestamp: 'desc' },
      take: limit,
    });
    
    return messages.reverse().map(message => new ChatMessage(message));
  }

  static async getBySessionId(sessionId) {
    const messages = await prisma.chatMessage.findMany({
      where: { sessionId },
      orderBy: { timestamp: 'asc' },
    });
    
    return messages.map(message => new ChatMessage(message));
  }

  static async clearHistory() {
    await prisma.chatMessage.deleteMany();
  }

  static async delete(id) {
    await prisma.chatMessage.delete({
      where: { id },
    });
  }

  toJSON() {
    return {
      id: this.id,
      content: this.content,
      type: this.type,
      timestamp: this.timestamp,
      context: this.context,
      isError: this.isError,
      isDiagnostic: this.isDiagnostic,
      diagnosticQuestions: this.diagnosticQuestions,
      diagnosticAnswers: this.diagnosticAnswers,
      parentMessageId: this.parentMessageId,
      createdAt: this.createdAt
    };
  }
}

module.exports = ChatMessage;

File: luma/lib/models/chat_message.dart

enum MessageType {
  user,
  assistant,
}

class ChatMessage {
  final String id;
  final String content;
  final MessageType type;
  final DateTime timestamp;
  final String? context; // Additional context like current week
  final bool isError;
  final bool isDiagnostic;
  final List<String>? diagnosticQuestions;
  final Map<String, dynamic>? diagnosticAnswers;
  final String? parentMessageId;
  final String? sessionId;

  ChatMessage({
    required this.id,
    required this.content,
    required this.type,
    required this.timestamp,
    this.context,
    this.isError = false,
    this.isDiagnostic = false,
    this.diagnosticQuestions,
    this.diagnosticAnswers,
    this.parentMessageId,
    this.sessionId,
  });

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'content': content,
      'type': type.name,
      'timestamp': timestamp.toIso8601String(),
      'context': context,
      'isError': isError,
      'isDiagnostic': isDiagnostic,
      'diagnosticQuestions': diagnosticQuestions,
      'diagnosticAnswers': diagnosticAnswers,
      'parentMessageId': parentMessageId,
      'sessionId': sessionId,
    };
  }

  factory ChatMessage.fromJson(Map<String, dynamic> json) {
    return ChatMessage(
      id: json['id'] ?? '',
      content: json['content'] ?? '',
      type: MessageType.values.firstWhere(
        (e) => e.name == json['type'],
        orElse: () => MessageType.user,
      ),
      timestamp: DateTime.parse(json['timestamp']),
      context: json['context']?.toString(),
      isError: json['isError'] ?? false,
      isDiagnostic: json['isDiagnostic'] ?? false,
      diagnosticQuestions: json['diagnosticQuestions'] != null 
          ? List<String>.from(json['diagnosticQuestions']) 
          : null,
      diagnosticAnswers: json['diagnosticAnswers'],
      parentMessageId: json['parentMessageId']?.toString(),
      sessionId: json['sessionId']?.toString(),
    );
  }

  ChatMessage copyWith({
    String? id,
    String? content,
    MessageType? type,
    DateTime? timestamp,
    String? context,
    bool? isError,
    bool? isDiagnostic,
    List<String>? diagnosticQuestions,
    Map<String, dynamic>? diagnosticAnswers,
    String? parentMessageId,
    String? sessionId,
  }) {
    return ChatMessage(
      id: id ?? this.id,
      content: content ?? this.content,
      type: type ?? this.type,
      timestamp: timestamp ?? this.timestamp,
      context: context ?? this.context,
      isError: isError ?? this.isError,
      isDiagnostic: isDiagnostic ?? this.isDiagnostic,
      diagnosticQuestions: diagnosticQuestions ?? this.diagnosticQuestions,
      diagnosticAnswers: diagnosticAnswers ?? this.diagnosticAnswers,
      parentMessageId: parentMessageId ?? this.parentMessageId,
      sessionId: sessionId ?? this.sessionId,
    );
  }
}
